File: src/command/command.parser.ts
import type { CommandContext, ParsedCommand } from './command.types.js';

export class CommandParser {
  private static readonly PREFIX = '!';

  parse(input: string): CommandContext | null {
    const trimmed = input.trim();

    if (!this.isCommand(trimmed)) {
      return null;
    }

    const withoutPrefix = this.removePrefix(trimmed);

    if (this.isEmpty(withoutPrefix)) {
      return null;
    }

    const parts = this.splitInput(withoutPrefix);

    return {
      raw: input,
      ...parts,
    };
  }

  private isCommand(input: string): boolean {
    return input.startsWith(CommandParser.PREFIX);
  }

  private removePrefix(input: string): string {
    return input.slice(CommandParser.PREFIX.length).trim();
  }

  private isEmpty(input: string): boolean {
    return input.length === 0;
  }

  private splitInput(input: string): ParsedCommand {
    const parts = input.split(/\s+/).filter(Boolean);

    return {
      command: parts[0],
      args: parts.slice(1),
    };
  }
}

File: Mythlog/Parser/CommandParser.cs
using Mythlog.Command;

namespace Mythlog.Parser;

public class CommandParser : ICommandParser
{
    private const string Prefix = "!";

    public CommandContext? Parse(string input)
    {
        if (string.IsNullOrWhiteSpace(input))
        {
            return null;
        }

        var trimmed = input.Trim();

        if (!IsCommand(trimmed))
        {
            return null;
        }

        var withoutPrefix = RemovePrefix(trimmed);

        if (IsEmpty(withoutPrefix))
        {
            return null;
        }

        var parts = SplitInput(withoutPrefix);

        return new CommandContext(Raw: input, Command: parts.Command, Args: parts.Args);
    }

    private static bool IsCommand(string input)
    {
        return input.StartsWith(Prefix);
    }

    private static string RemovePrefix(string input)
    {
        return input[Prefix.Length..].Trim();
    }

    private static bool IsEmpty(string input)
    {
        return input.Length == 0;
    }

    private static ParsedCommand SplitInput(string input)
    {
        var parts = input.Split((char[])null!, StringSplitOptions.RemoveEmptyEntries);
        return new ParsedCommand(Command: parts[0], Args: parts[1..].ToArray());
    }
}